// member_integral/lib/integral.js

module.exports = function(req) {
//	  req.app.db.models.Account ,req.user.roles.account.integral;
function Integral(){
	this.columnExist();
}
/**
 * 是否存在积分字段判断
 */
Integral.prototype.columnExist = function(){
	
	if(!req.user.roles.account.integral){
		var fieldsToSet = {
				points : 0,
				level : 1,
				levelName : '普通会员',
				consumeMoney : 0,
		};
		req.user.roles.account.integral = fieldsToSet;
	}
};
/**
 * 总添加
 */
Integral.prototype.convertCode = function(addMoney ,callback){
	if(addMoney == ''){
		callback('addMoney is necessary');
	}
	var pattern = /^-?([1-9]\d*|[1-9]\d*\.\d*|0\.\d*[1-9]\d*)$/; 
	var flag = pattern.test(addMoney);
	if(!flag){
		callback('addMoney is Invalid');
	}
	var addMoney = Number(addMoney);
	//获取活动方案
	var conditions = {
			isUse :'yes',
	};
	//查找积分兑换比率模板
	req.app.db.models.Integral.findOne(conditions , function(err ,integral){
		if(err){
			callback(err);
		}
		if(integral){
			
			//根据等级获取积分比率计算积分
			var PointsPercent = eval('integral.PointsPercentOfLeve.lv'+req.user.roles.account.integral.level);
			var addPoints = Math.round(PointsPercent.percent * addMoney);
			var newPoints = req.user.roles.account.integral.points + addPoints;
			
			//添加金额
			var newConsumeMoney = req.user.roles.account.integral.consumeMoney + addMoney;
			
			//计算等级
				//当前等级下是否还有等级
			var level = req.user.roles.account.integral.level;
			
			var addLevel = 0;
			var newLevel = 0;
			var newLevelName = req.user.roles.account.integral.levelName;
			
			while(true){
				PointsPercent = eval('integral.PointsPercentOfLeve.lv' + level);
				
				if( PointsPercent && ( newConsumeMoney >= PointsPercent.consumeMoney ) ){
					addLevel = level - req.user.roles.account.integral.level;
					newLevel = level;
					newLevelName = PointsPercent.levelName;
					
				}else{
					break;
				}
				//等级+1
				level++;
			}
			//更新
			//设置方式
			var fieldsToSet = {
					'integral.consumeMoney' : newConsumeMoney,
					'integral.points' : newPoints,
					'integral.levelName' : newLevelName,
					'integral.level' : newLevel,
			};
			
			
			req.app.db.models.Account.findByIdAndUpdate(req.user.roles.account.id ,fieldsToSet ,function(err ,queryObj){
				if(err){
					return callback(err);
				}
				
				if(queryObj){
					//***************此处为返回信息,可用于日志记录**************************
					var result = {
							//增加之前的信息
							oldConsumeMoney :req.user.roles.account.integral.consumeMoney,
							oldPoints :req.user.roles.account.integral.points,
							oldLevelName :req.user.roles.account.integral.levelName,
							oldLevel :req.user.roles.account.integral.level,
							//增加信息
							addMoney : addMoney,
							addPoints: addPoints,
							addLevel : addLevel,
					};
					//增加后的信息
					req.user.roles.account.integral.consumeMoney = newConsumeMoney;
					req.user.roles.account.integral.points = newPoints;
					req.user.roles.account.integral.levelName = newLevelName;
					req.user.roles.account.integral.level = newLevel;
					
					callback(null ,result);
				}
			});
		}
		else{
			callback('err : Can\'t find integral');
		}
	});
};
/**
 * 金额
 */

/**
 * 等级
 */

/**
 * 积分
 */
//单纯积分增加
Integral.prototype.addPoints = function(addPoints ,callback){
	if(addPoints == ''){
		callback('addPoints is necessary');
	}
	var pattern = /^-?([1-9]\d*|[1-9]\d*\.\d*|0\.\d*[1-9]\d*)$/; 
	var flag = pattern.test(addPoints);
	if(!flag){
		callback('addPoints is Invalid');
	}
	var addPoints = Number(addPoints);
	
	var newPoints = req.user.roles.account.integral.points + addPoints;
	
	var fieldsToSet = {
			'integral.points' : newPoints,
	};
	
	
	req.app.db.models.Account.findByIdAndUpdate(req.user.roles.account.id ,fieldsToSet ,function(err ,queryObj){
		if(err){
			callback(err);
		}
		
		if(queryObj){
			//***************此处为返回信息,可用于日志记录**************************
			var result = {
					//增加之前的信息
					oldPoints :req.user.roles.account.integral.points,
					//增加信息
					addPoints: addPoints,
			};
			//增加后的信息
			req.user.roles.account.integral.points = newPoints;
			
			callback(null ,result);
		}
	});
};

//单纯积分减少
Integral.prototype.reducePoints = function(reducePoints ,callback){
	if(reducePoints == ''){
		callback('reducePoints is necessary');
	}
	var pattern = /^-?([1-9]\d*|[1-9]\d*\.\d*|0\.\d*[1-9]\d*)$/; 
	var flag = pattern.test(reducePoints);
	if(!flag){
		callback('reducePoints is Invalid');
	}
	var reducePoints = Number(reducePoints);
	
	var newPoints = req.user.roles.account.integral.points - reducePoints;
		
		var fieldsToSet = {
				'integral.points' : newPoints,
		};
		
		
		req.app.db.models.Account.findByIdAndUpdate(req.user.roles.account.id ,fieldsToSet ,function(err ,queryObj){
			if(err){
				callback(err);
			}
			
			if(queryObj){
				//***************此处为返回信息,可用于日志记录**************************
				var result = {
						//增加之前的信息
						oldPoints :req.user.roles.account.integral.points,
						//增加信息
						reducePoints: reducePoints,
				};
				//增加后的信息
				req.user.roles.account.integral.points = newPoints;
				
				callback(null ,result);
			}
		});
};
/**
 * 金币
 */

return Integral;
};
//module.exports = Integral;
