// weixin/lib/weixin.js
var https = require('https');
/**
 * 微信模块 用于获取用户ipenid 和多个公众号openid和账户之间的关联关系
 * 
 * 流程:
 * 1.微信用户通过分享页面进入(不从服务号进入),默认来源是此appid的来源
 * 2.微信用户通过公众号外链进入,以get的方式带着openid和来源
 * 
 * @author zhongxinwen
 * @returns
 */

function Weixin(){
	this.appID = 'wxc0201826fb70f56a'||null;
	this.appsecret = '6aafe0416674951851123105ff54033d'||null;
	this.webUrls = {
			getCode:'https://open.weixin.qq.com/connect/oauth2/authorize',
			getAccessToken:'https://api.weixin.qq.com/sns/oauth2/access_token',
			refresAccessToken:'https://api.weixin.qq.com/sns/oauth2/refresh_token',
			getUserinfo:'https://api.weixin.qq.com/sns/userinfo',
	};
	//本机微信回调页面
	this.callback = "";
}

Weixin.prototype.init = function(){
	
};
//发送请求
//发送request
Weixin.prototype.sendRequest = function(url , callback){
	https.get(url, function(ress) {
		  ress.setEncoding('utf8');
		  var body='';
		　　ress.on('data', function(data) {
				body += data;
		　　}).on('end', function () {
				body = JSON.parse(body);
				return callback(null ,body);
			});
		}).on('error', function(e) {
		  console.log("Got error: " + e.message);
		  return callback("Got error: " + e.message);
		});
};

//网页授权
//第一步 生成这个的回调页面 获取code
/*
 * https://open.weixin.qq.com/connect/oauth2/authorize?
 * appid=APPID&
 * redirect_uri=REDIRECT_URI&			//这个需要带有 http
 * response_type=code&
 * scope=SCOPE&
 * state=STATE
 * #wechat_redirect
*/

Weixin.prototype.callbackUrl = function (option){
	var callbackurl = option.callbackurl || '';
	//这个可以存放来源的简称,便于统计(后期可以作为授权第三方的参数签名认证)
	var state = option.state || 'ali_member_path';
	
	var params = 'appid=' + this.appID + '&redirect_uri='+ encodeURIComponent(callbackurl) +'&response_type=code&scope=snsapi_base&state='+state+'#wechat_redirect';
	var callbackurl = this.webUrls.getCode + '?' + params;
	
	return callbackurl;
	
};
//获取到code 然后进行验证获取accesstoken 和openid (base模式下)
/*https://api.weixin.qq.com/sns/oauth2/access_token?
 * appid=APPID&
 * secret=SECRET&
 * grant_type=authorization_code
 * */
Weixin.prototype.webGrant = function (code ,callback){
	var params = 'appid='+this.appID+'&secret='+this.appsecret+'&code='+code+'&grant_type=authorization_code';
	var requestUrls = this.webUrls.getAccessToken + '?' + params;
	this.sendRequest(requestUrls , function(err ,data){
		if(err){
			return callback(err);
		}
		if(data){
			/*
			 * {
				   "access_token":"ACCESS_TOKEN",
				   "expires_in":7200,
				   "refresh_token":"REFRESH_TOKEN",
				   "openid":"OPENID",
				   "scope":"SCOPE"
				}
			 */
			callback(null ,data);
		}else{
			callback('err : unfound openid!');
		}
		
	});
};
	
Weixin.prototype.mergeOpenid = function(req, user, callback){
	//将查询到的user openid 循环添加到关联帐号
	//关联到的账户的openid
	var userLenth = user.weixin.openid.length;
	 
	//临时帐号的openid
	var tmpUserLength =  req.user.weixin.openid.length
    
	//循环对比
	 for(var i=0 ;i < tmpUserLength ;i++){
		 var isExist = false;
		 for(var j=0 ; j < userLenth ;j++){
			if(user.weixin.openid[j] == req.user.weixin.openid[i]){
				isExist = true;
				break;
			}
	     }
	}
	//不存在 填入user.weixin.openid 数组里面
	if(!isExist){
		user.weixin.openid.push(req.user.weixin.openid[i]);
	}
	
	if(userLenth < user.weixin.openid.length){
    	//存入openid并且更新
		var fieldsToSet = {
				'weixin.openid' :user.weixin.openid,
		};
		console.log('临时帐号openid');
		console.log(req.user.weixin.openid);
		console.log('关联帐号openid');
		console.log(user.weixin.openid);
		
		console.log('更新关联帐号的openid');
		
		req.app.db.models.User.findByIdAndUpdate( user._id ,fieldsToSet ,function(err ,queryObj){
			if(err){
				return callback(err);
			}
			//更新成功,存入session
			if(queryObj){
				console.log('更新成功,开始删除临时帐号');
				//删除临时user account
				console.log(req.user);
				//等级新session
				console.log('使用关联帐号登录');
				return callback(null ,user);
			}else{
				console.log('关联失败');
				return callback('更新用户失败');
			}
		});
		
	}else{
		return callback(null,user);
	}
};

module.exports = Weixin;
