// passport-ali_discuz/lib/strategy.js
var http = require('http')
	,sha1 = require('./sha1');

function Strategy() {
 this.localhost = null;
 this.urlhost = null;
 this.urls={
		 authorize:'/api/interface/index.php?module=authorize&action=login&callbak=',
		 getuserinfo:'/api/interface/index.php?module=user&action=getuserinfo&token=',
 };
 this.source = 'ali_member_pathform';
}

Strategy.prototype.init = function(urlhost){
	return require('./init')(this ,urlhost);
}

Strategy.prototype.sethost = function(localhost ,urlhost){
	if(localhost.substr(-1)=='/'){
		localhost = localhost.substr(0 ,(localhost.length-1));
	}
	if(urlhost.substr(-1)=='/'){
		urlhost = urlhost.substr(0 ,(urlhost.length-1));
	}
	this.localhost = localhost;
	this.urlhost = urlhost;
}


// 发送request
Strategy.prototype.sendRequest = function(url , callback){
	http.get(url, function(ress) {
		  ress.setEncoding('utf8');
		  var body='';
		　　ress.on('data', function(data) {
				body += data;
		　　}).on('end', function () {
				body = JSON.parse(body);
				if(body.statusCode == 1){
					callback(null ,body.data);
				}else{
					callback("Got error: " + JSON.stringify(body.error)+",message: "+JSON.stringify(body.msg));
				}
			});
		}).on('error', function(e) {
		  console.log("Got error: " + e.message);
		  callback("Got error: " + e.message);
		});
};

//login url
Strategy.prototype.loginUrl = function(option){
	option.callbak = this.localhost + option.callbak;
	return this.urlhost + this.urls.authorize + encodeURIComponent(option.callbak) + this.urlParams(option);
};

//params
Strategy.prototype.urlParams = function(option){
	var noce = require('./ranstr')(32);
	var noce_encode = encodeURIComponent(noce);
	var time = Date.now();
	var sigString_base64 ='';
	
	if(option.accesstoken){
//		token+time+noce(解base64)
		sigString_encode = encodeURIComponent(sha1.HMACSHA1( this.source ,option.accesstoken + time + noce));
	}
	else if(option.callbak){
//		callback(解base64)+time+noce(解base64)
		sigString_encode = encodeURIComponent(sha1.HMACSHA1( this.source ,option.callbak + time + noce));
	}
//	console.log(this.source);
//	console.log(option.callbak + time + noce);
//	console.log(sigString);
	return '&source=' + this.source + '&time=' + time +'&noce=' + noce_encode + '&signature='+sigString_encode;
};

// authenticate and get userinfo
Strategy.prototype.authenticate = function(accesstoken , callback){
	var option = {accesstoken:accesstoken};
	var url = this.urlhost +this.urls.getuserinfo + accesstoken +this.urlParams(option);

	this.sendRequest(url ,function(err ,data){
		if(err){
			return callback(err);
		}
		if(data){
			var profile = { provider: 'ali_discuz' };
			profile.id = data.uid;
			profile.displayName = '';
			profile.username = data.username;
			profile.profileUrl = '';
			profile.emails = data.email;
			
			profile._raw = JSON.stringify(data);
			profile._json = data;
			callback(null ,profile);
		}
		else{
			callback("Got error: " +'user infomations is not found!');
		}
	});
};
module.exports = Strategy;
